coral-common/src/main/java/com/linkedin/coral/common/IcebergTypeConverter.java

```diff
@@ -0,0 +1,218 @@
+    List<RelDataType> fieldTypes = new ArrayList<>(columns.size());
+    List<String> fieldNames = new ArrayList<>(columns.size());
+
+    for (Types.NestedField field : columns) {
+      fieldNames.add(field.name());
+
+      // Convert field type using the main dispatcher
+      RelDataType fieldType = convert(field.type(), typeFactory);
+
+      // Handle nullability - Iceberg has explicit required/optional
+      fieldType = typeFactory.createTypeWithNullability(fieldType, field.isOptional());
+
+      fieldTypes.add(fieldType);
+    }
+
+    return typeFactory.createStructType(fieldTypes, fieldNames);
+  }
+
+  /**
+   * Main dispatcher - converts Iceberg Type to Calcite RelDataType based on type category.
+   * Similar to TypeConverter.convert(TypeInfo, RelDataTypeFactory).
+   * 
+   * @param icebergType Iceberg type
+   * @param typeFactory Calcite type factory
+   * @return RelDataType representing the Iceberg type
+   */
+  public static RelDataType convert(Type icebergType, RelDataTypeFactory typeFactory) {
+    Type.TypeID typeId = icebergType.typeId();
+
+    switch (typeId) {
+      case STRUCT:
+        return convert((Types.StructType) icebergType, typeFactory);
+      case LIST:
+        return convert((Types.ListType) icebergType, typeFactory);
+      case MAP:
+        return convert((Types.MapType) icebergType, typeFactory);
+      default:
+        // Handle all primitive types
+        return convertPrimitive(icebergType, typeFactory);
+    }
+  }
+
+  /**
+   * Converts Iceberg primitive types to Calcite RelDataType.
+   * Handles all atomic types: BOOLEAN, INTEGER, LONG, FLOAT, DOUBLE, DATE, TIME, TIMESTAMP,
+   * STRING, UUID, FIXED, BINARY, DECIMAL.
+   * 
+   * @param icebergType Iceberg primitive type
+   * @param typeFactory Calcite type factory
+   * @return RelDataType representing the primitive type
+   */
+  public static RelDataType convertPrimitive(Type icebergType, RelDataTypeFactory typeFactory) {
+    RelDataType convertedType = null;
+    Type.TypeID typeId = icebergType.typeId();
+
+    switch (typeId) {
+      case BOOLEAN:
+        convertedType = typeFactory.createSqlType(SqlTypeName.BOOLEAN);
+        break;
+      case INTEGER:
+        convertedType = typeFactory.createSqlType(SqlTypeName.INTEGER);
+        break;
+      case LONG:
+        convertedType = typeFactory.createSqlType(SqlTypeName.BIGINT);
+        break;
+      case FLOAT:
+        convertedType = typeFactory.createSqlType(SqlTypeName.FLOAT);
+        break;
+      case DOUBLE:
+        convertedType = typeFactory.createSqlType(SqlTypeName.DOUBLE);
+        break;
+      case DATE:
+        convertedType = typeFactory.createSqlType(SqlTypeName.DATE);
+        break;
+      case TIME:
+        convertedType = typeFactory.createSqlType(SqlTypeName.TIME);
+        break;
+      case TIMESTAMP:
+        // Iceberg timestamp type - microsecond precision (6 digits)
+        convertedType = typeFactory.createSqlType(SqlTypeName.TIMESTAMP, 6);
```

Comment:
Can we handle timestamp with time zone for completeness?
```suggestion
        Types.TimestampType timestampType = (Types.TimestampType) icebergType;
        if (timestampType.shouldAdjustToUTC()) {
                // TIMESTAMP WITH TIME ZONE - stores instant in time
               convertedType = typeFactory.createSqlType(SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE, 6);
        } else {
               // TIMESTAMP - stores local datetime
               convertedType = typeFactory.createSqlType(SqlTypeName.TIMESTAMP, 6);
        }
```
Ref: https://github.com/apache/calcite/blob/95350ed1a449bbb2f008fcf2b704544e7d95c410/core/src/main/java/org/apache/calcite/sql/type/SqlTypeName.java#L73